src/storage/src/source/generator/key_value.rs

```diff
@@ -0,0 +1,584 @@
+                            key_value.keys,
+                            key_value.snapshot_rounds,
+                            key_value.quick_rounds,
+                            key_value.batch_size,
+                            key_value.seed,
+                            upper_offset,
+                            key_value.include_offset.is_some(),
+                        ),
+                    )
+                })
+            })
+            .collect();
+        if !local_partitions.is_empty()
+            && (key_value.update_rate.is_some() || key_value.quick_rounds > 0)
+        {
+            let mut interval = key_value.update_rate.map(tokio::time::interval);
+
+            loop {
+                if local_partitions.values().all(|si| si.finished_quick()) {
+                    if let Some(interval) = &mut interval {
+                        interval.tick().await;
+                    } else {
+                        break;
+                    }
+                }
+
+                for up in local_partitions.values_mut() {
+                    updates_buffer.clear();
+                    upper_offset = up.produce_batch(&mut updates_buffer, &mut value_buffer);
+                    data_output.give_container(&cap, &mut updates_buffer).await;
+                }
+                cap.downgrade(&MzOffset::from(upper_offset));
+                progress_cap.downgrade(&MzOffset::from(upper_offset));
+                tokio::task::yield_now().await;
+            }
+        }
+
+        std::future::pending::<()>().await;
+    });
+
+    let status = [HealthStatusMessage {
+        index: 0,
+        namespace: StatusNamespace::Generator,
+        update: HealthStatusUpdate::running(),
+    }]
+    .to_stream(scope);
+    let stats_stream = stats_stream.concat(&steady_state_stats_stream);
+
+    (
+        stream.as_collection(),
+        Some(progress_stream),
+        status,
+        stats_stream,
+        vec![button.press_on_drop(), stats_button],
+    )
+}
+
+/// An iterator that produces keys belonging to a partition.
+struct PartitionKeyIterator {
+    // The partition.
+    partition: u64,
+    // The number of partitions.
+    partitions: u64,
+    // The key space.
+    keys: u64,
+    // The next key.
+    next: u64,
+}
+
+impl PartitionKeyIterator {
+    fn new(partition: u64, partitions: u64, keys: u64, start_key: u64) -> Self {
+        assert_eq!(keys % partitions, 0);
+        PartitionKeyIterator {
+            partition,
+            partitions,
+            keys,
+            next: start_key,
+        }
+    }
+}
+
+impl Iterator for &mut PartitionKeyIterator {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ret = self.next;
+        self.next = (self.next + self.partitions) % self.keys;
+        Some(ret)
+    }
+}
+
+/// A struct that produces batches of data for the snapshotting phase.
+struct SnapshotProducer {
+    // The key iterator for the partition.
```

Comment:
doc comments here too 
```suggestion
    /// The key iterator for the partition.
```